main/src/library/MutHeap.flix

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright 2023 Xavier deSouza
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+pub enum MutHeap[a: Type, r: Region] {
+    case MutHeap(Region[r], Ref[Array[a, r], r], Ref[Int32, r])
+}
+
+mod MutHeap {
+
+    ///
+    /// Returns an empty MutHeap
+    ///
+    pub def new(rc: Region[r]): MutHeap[a, r] \ r =
+        MutHeap(rc, ref Array.new(rc, 8) @ rc, ref 0 @ rc)
+
+    ///
+    /// Returns an empty MutHeap with a reserved amount of space
+    ///
+    pub def newWithSpace(rc: Region[r], size: Int32): MutHeap[a, r] \ r =
+        MutHeap(rc, ref Array.new(rc, size) @ rc, ref 0 @ rc)
+
+    ///
+    /// Reserves a specific amount of space for
+    ///
+    pub def reserve(h: MutHeap[a, r], size: Int32): Unit \ r =
+        let MutHeap(rc, arr, s) = h;
+        if (size > deref s) {
```

Comment:
This should probably check capacity rather than size. You could also make a `capacity` helper function to make this clearer.